google/cloud/universe_domain_test.cc

```diff
@@ -0,0 +1,122 @@
+    "client_email": "foo-email@foo-project.iam.gserviceaccount.com",
+    "client_id": "100000000000000000001",
+    "auth_uri": "https://accounts.google.com/o/oauth2/auth",
+    "token_uri": "https://accounts.google.com/o/oauth2/token",
+    "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
+    "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/foo-email%40foo-project.iam.gserviceaccount.com",
+    "universe_domain": "test-ud.net"
+})""";
+
+/// @test ServiceAccount credentials support universe_domain and should
+/// return the universe_domain, if present.
+TEST(AddUniverseDomainOption, GoogleDefaultCredentialsServiceAccount) {
+  auto const filename = TempFileName();
+  std::ofstream(filename) << kServiceAccountCredContents;
+  auto const env =
+      ScopedEnvironment(oauth2_internal::GoogleAdcEnvVar(), filename.c_str());
+  auto result_options = AddUniverseDomainOption(ExperimentalTag{}, Options{});
+  (void)std::remove(filename.c_str());
+
+  ASSERT_STATUS_OK(result_options);
+  ASSERT_TRUE(result_options->has<UnifiedCredentialsOption>());
+  auto creds = result_options->get<UnifiedCredentialsOption>();
+  EXPECT_THAT(creds.get(), NotNull());
+  ASSERT_TRUE(result_options->has<internal::UniverseDomainOption>());
+  EXPECT_THAT(result_options->get<internal::UniverseDomainOption>(),
+              Eq("test-ud.net"));
+}
+
+TEST(AddUniverseDomainOption, CredentialsSpecified) {
+  auto expected_creds = MakeInsecureCredentials();
+  auto options = Options{}.set<UnifiedCredentialsOption>(expected_creds);
+  auto result_options = AddUniverseDomainOption(ExperimentalTag{}, options);
+
+  ASSERT_STATUS_OK(result_options);
+  ASSERT_TRUE(result_options->has<UnifiedCredentialsOption>());
+  auto const& creds = result_options->get<UnifiedCredentialsOption>();
+  // Verify UnifiedCredentialsOption value did not get overwritten.
+  EXPECT_THAT(expected_creds.get(), Eq(creds.get()));
```

Comment:
nit: I don't think we need to be `.get()`ting the raw pointer in any of these tests. We can set expectations on the shared_ptr.